### PR TITLE
Unify code to fill/empty buckets, Part 2

### PR DESCRIPTION
**Pull request by HenryLoenwind**
* New: Emptying buckets
* Fix: Fixed many bugs, including some item-loss types
* Full IN+OUT: Tank, Reservoir, XP Obelisk, Soul Binder
* IN for input liquid, OUT for output liquid: Vat
* IN for fuel, OUT fro XP: KillerJoe
* IN for fuel: Zombie Generator
* IN for fuel/coolant: Combustion Generator
* re #1913

### CODE REVIEW

**Comment by HenryLoenwind** [comment]
This turned from a small change into a much bigger rewrite after a tiny bug was found. So this is WIP, sorry.

**Comment by keturn** [comment]
I found this pull request when looking in to #1913.

I think factoring out `fillInternalTankFromPlayerHandItem` from the various `onBlockActivated` methods is a good change, and `tryDrainContainer` looks like it'll handle more cases than the old `Util.consumeItem` did. (I'm assuming the TE items like Portable Tank and Sponge implement IFluidContainerItem?)

**Comment by HenryLoenwind** [comment]
@keturn Yes, I think they do. I'll have to do more tests with other tanks later. Currently tested are ExU's Drums (ok), Mariculture's Ladle (ok) and Tank (item form doesn't work), Vanilla Bucket and Bottle (ok).

**Comment by HenryLoenwind** [comment]
@tterrag1098 This is done now

**Comment by tterrag1098** [comment]
I have a feeling this wasn't properly tested...

File: src/main/java/crazypants/enderio/BlockEio.java [hunk 1]
```diff
@@ -91,6 +93,14 @@ public boolean onBlockActivated(World world, int x, int y, int z, EntityPlayer e
         world.markBlockForUpdate(x, y, z);
         return true;
       }
+      if (te instanceof ITankAccess) {
```
Comment (tterrag1098) [inline src/main/java/crazypants/enderio/BlockEio.java]: This code is inside a check for tools...so it never functions

File: src/main/java/crazypants/util/FluidUtil.java [hunk 2]
```diff
@@ -227,47 +328,137 @@ public static ItemStack fillContainerFromInternalTank(ITankAccess tank, ItemStac
...
+    if (!entityPlayer.capabilities.isCreativeMode) {
+      if (fill.remainder.itemStack == null) {
+        entityPlayer.inventory.setInventorySlotContents(entityPlayer.inventory.currentItem, fill.result.itemStack);
+        return true;
+      } else {
+        entityPlayer.inventory.setInventorySlotContents(entityPlayer.inventory.currentItem, fill.remainder.itemStack);
+      }
+    }
+
+    if (fill.result.itemStack.isStackable()) {
```
Comment (tterrag1098) [inline src/main/java/crazypants/util/FluidUtil.java]: This code being outside the creative mode check makes buckets spawn for no reason

File: src/main/java/crazypants/util/FluidUtil.java [hunk 3]
```diff
@@ -227,47 +328,137 @@ public static ItemStack fillContainerFromInternalTank(ITankAccess tank, ItemStac
...
-        item.stackSize--;
-        entityPlayer.inventory.setInventorySlotContents(entityPlayer.inventory.currentItem, item);
-      }
-      if (filledContainer.isStackable()) {
+    for (FluidTank subTank : tank.getOutputTanks()) {
+      FluidAndStackResult fill = tryFillContainer(entityPlayer.inventory.getCurrentItem(), subTank.getFluid());
+      if (fill.result.fluidStack != null) {
+
+        if (!entityPlayer.capabilities.isCreativeMode) {
+          subTank.setFluid(fill.remainder.fluidStack);
```
Comment (tterrag1098) [inline src/main/java/crazypants/util/FluidUtil.java]: This code being *inside* the creative check makes filling buckets not work at all in creative

**Comment by tterrag1098** [comment]
I realize this is a large patch but after literally 30 seconds of testing I was able to figure out all of these and fix them

File: src/main/java/crazypants/enderio/BlockEio.java (hunk 1 above)
Comment (HenryLoenwind) [inline src/main/java/crazypants/enderio/BlockEio.java]: To be fair, this looked just like all those getCurrentItem()s...

File: src/main/java/crazypants/util/FluidUtil.java (hunk 3 above)
Comment (HenryLoenwind) [inline src/main/java/crazypants/util/FluidUtil.java]: It does, the creative-check is around "remove fluid from tank" and "remove empty bucket from player".

**Pull request merged**